### ISSUE TITLE
Need to prepare release

### ISSUE THREAD

**Issue by pemensik** [raceintospace/raceintospace#416]
Game code have received too many modifications since last release. And yet, no new release was done.

We should prepare TODO list and prepare release ASAP. If we are unable to fix some errors, just postpone them to another release.

Even portability changes deserves new release. If we want packages with recent modifications, we should make some release packagers can reuse. There is nothing usable right now. I have used snapshot on Fedora, but release would be better.

My question is, what is missing? Are we able to prepare Windows and MacOS installer?

**Comment by peyre** [comment]
MacOS we may not be able to swing this time around.  I saw the guys were working on it, but last I heard it wouldn't compile on the Mac nowadays.

For Windows we might need to consult with Ryan, and maybe Krzysztof Kościuszkiewicz, who used to prepare the Windows releases when we were on SourceForge.  I'd be glad to help too, but I don't know how (but I would like to learn!).

For Linux, Pace Willisson says he'd be happy to come back and take care of them.

**Comment by peyre** [comment]
Incidentally, over the weekend I created replacement videos for the N1 and UR-700 rockets, and for the Lapot minishuttle, and have started testing them in the game.  (This is per #112.)  I tried this once before with the rocket videos being a mod you could copy and paste into an installed copy of RIS.  It worked on mine, which was compiled from source on Linux, but in Windows the movie area of the screen showed a solid green.  Now with these, I've noticed that if I copy and paste them into the game after compiling, the old movies play.  If I copy them in and then compile, they work, and so far they're working.

Assuming these replacement videos work out, I'd like to include them in the next release.  The Lapot videos are more footage of the M2-F2 etc. and I've replaced them with footage of actual MiG-105 Spiral drop tests.  The rocket videos are crude copy-and-pastes of the rockets into videos of other rocket launches, and I've replaced them with actual N1 launch footage and footage of the Energiya-Polyus, which was based on the UR-700.

**Comment by peyre** [comment]
Ah, no.  I did some testing last night, and it seems the movies I spent the weekend making...simply don't work in the game.  It looks like the ones I made some years ago _do_ work, but they're not nearly as good: they have watermarks, etc.  So we may have to skip the mission movies in the next release.

**Comment by rnyoakum** [comment]
Some thoughts about this:

- @pemensik is right. I think the project has been stalled on releases in large part because Mac compliance isn't there, but... who knows when that's going to happen?
- Because RIS generally puts all updates on the master, there's no need to create any release thread; just tag the release commit;
- however, there should be a Development branch created so contributions can be added while preparing for release. Any work involved in finalizing the release should be done on the master branch (possibly mirrored to the Development branch). After releasing, the Development branch can be merged into master.
- Any bugs that need fixing before release should be tracked with the [milestone](https://github.com/raceintospace/raceintospace/milestone/2). 


As for the new features:
> I did some testing last night, and it seems the movies I spent the weekend making...simply don't work in the game.
> @peyre

I recommend including any new movies that may or may not work in a future commit. A release should be as bug-free as possible. There will still be issues, but it's best not to create potential new ones in the release.

My inclination would be to draw a hard line. No more new features, only **simple** bug fixes on the master until the next release - and then plan for when the next release should be.

Except... that brings up the issue of @hweimer's PBEM commit https://github.com/raceintospace/raceintospace/pull/419. It appears to work but
> At some points, I have hit some ancient codepaths, like when starting the mail game by calling Prefs(3). While things seem to work, there might be unintended side effects. Therefore, expect bugs.

Should this be fit into the release? It's a useful feature, and I'm while I'm inclined to say _no, it's too soon_ because
- It's 600+ lines of changed code restoring a feature that handle major elements of the game differently. There may be a number of insidious bugs, and there would be little time to address them before release.
- https://github.com/raceintospace/raceintospace/pull/419 changes the size of the game data, which invalidates previous save files.
- If changing the data written to the save file, there are other changes that should be done at the same time - it's something best done infrequently, and there are several issues I haven't addressed specifically because of this point.

I'm thinking _yes_ because:
- PBEM was in the last release (I think).
- Any bugs introduced should be localized to PBEM.
- Having PBEM in the wild would help _identify_ any bugs, from having a larger test base.
- If updating to a new 2.0 version, that's the best time to change the save format, and how many people (other than me :) have save game files more than a year old?
- Having to release an update to fix any PBEM issues would provide an incentive to have a more regular release schedule.
- PBEM is a big deal.


If including PBEM in the release, it should be accompanied by a note that "_Play By E-mail had to be rewritten due to library availability issues and may be buggy_" - the audience deserves fair warning.

Thoughts?

**Comment by peyre** [comment]
Re. new movies: agreed, they should stay out.  I can try to try to sort them out later.

Re. PBEM: I'm inclined to say yes please, let's include it if we can.  I'd be glad to help playtest it.

Sadly, PBEM wasn't in the previous release.  The code for it was disabled in RIS, at least as far back as .4.5.  However they left the options for it (and Modem) on the main screen, but inactive.

I've wanted to bring back PBEM for a long time since it's the one big feature that BARIS had that RIS doesn't, plus it's the only way to play long-distance (barring sharing screens via Remote Desktop or something, anyway).

**Comment by hweimer** [comment]
I am somewhat hesitant towards including PBEM in the upcoming release. My expectation is that one will have to play several complete games to iron out the biggest issues and this will require quite some time. I also have some ideas for improvements, one being to introduce proper versioning for save games and thus retaining backwards compatibility. In my opinion, it is better to work on these things without rushing them. One can also turn one of Ryan's advantages around by saying that the inclusion of PBEM would warrant another release in the not-too-distant future and thereby also contributing towards a more regular release schedule.

**Comment by rnyoakum** [comment]
> I also have some ideas for improvements, one being to introduce proper versioning for save games and thus retaining backwards compatibility.
> @hweimer

I'd love to hear your thoughts on how to handle saves. The previous proposed solution (before I joined) was using protobuffs - I'm guessing you have a different approach? Also, let me know about any other improvements you're thinking about. If you don't want to hijack this thread, feel free to send me an email.

**Comment by peyre** [comment]
hweimer, you make some good arguments.  I would rather see the game released without PBEM than held up indefinitely because of it.  Whatever people decide is good with me.

**Comment by hweimer** [comment]
> @rnyoakum:
> I'd love to hear your thoughts on how to handle saves.

I was thinking about a minimally invasive solution. #419 adds to fields at the end of the Data struct, which are not being used at all in regular games. So, we can just populate the Data struct in these cases with what is found in an old save game. To distinguish new and old save games, we could use the currently unused Data->BUZZ, which is meant to be a "Save Version Marker" anyway.

> Also, let me know about any other improvements you're thinking about.

I was thinking about switching to an ABBAABB... ordering in PBEM games (i.e., like tie-breaks in tennis). After the first half-turn, the players could play two of their turns in a row, thus speeding up gameplay considerably. It would also put the U.S. and Soviet side on almost equal footing when it comes to being able to react to the other players Prestige Firsts and getting instantaneous feedback on mission prestige.

But you are perfectly right, the PR is already big enough, and the game logic behind PBEM games is also a bit easier to understand without meddling with the player order. So I'd rather work on that at some later point.

**Comment by rnyoakum** [comment]
> I am somewhat hesitant towards including PBEM in the upcoming release. My expectation is that one will have to play several complete games to iron out the biggest issues and this will require quite some time. I also have some ideas for improvements, one being to introduce proper versioning for save games and thus retaining backwards compatibility. In my opinion, it is better to work on these things without rushing them. One can also turn one of Ryan's advantages around by saying that the inclusion of PBEM would warrant another release in the not-too-distant future and thereby also contributing towards a more regular release schedule.

If this is your opinion, I'd be glad to create a Development branch and add your work onto it.

> I was thinking about switching to an ABBAABB... ordering in PBEM games (i.e., like tie-breaks in tennis)

Clever - I like it.

**Comment by pemensik** [comment]
I would vote against having development branch. On most projects I have seen, development is done on master branch.
Instead, releases are derived from master. If they support multiple releases, they make release branch, which may receive selected commits by git cherry-pick.

Regarding Play by Email, I think we can make it optional. It could be enabled or disabled on compile time. So it does not spoil release for those who wants it perfect. But to provide simple way to enable experimental features. Simple use of precompiler define, enabled by CMake boolean EXPERIMENTAL, might help.

**Comment by rnyoakum** [comment]
>I would vote against having development branch.

For the moment, there is a development branch I created. I did so last night before seeing your message, so please take no offense; I can fast-forward the master and remove development if that's the best course.  

>I would vote against having development branch. On most projects I have seen, development is done on master branch.
>Instead, releases are derived from master. If they support multiple releases, they make release branch, which may receive selected commits by git cherry-pick.

I'm not sure I'm following your statement, but what I think you're saying is to build up the release commit like so:

- Create a **release** branch off of **master**
- Development continues on **master**
- Select bug fixes & finishing touches are cherry-picked onto the **release** branch from **master**
- When ready, the latest commit on **release** is tagged and built
- Meanwhile, life goes on on **master**

If that's what you're suggesting, it's a sound approach and I'm fine with it instead. If not, would you help clear up my confusion?

> Regarding Play by Email, I think we can make it optional. It could be enabled or disabled on compile time. So it does not spoil release for those who wants it perfect. But to provide simple way to enable experimental features. Simple use of precompiler define, enabled by CMake boolean EXPERIMENTAL, might help.

This sounds perfectly reasonable.

**Comment by pemensik** [comment]
No problem with development branch. We can merge and remove it anytime.

I think it is not necessary for us to spawn release branch. I think better practice is to develop experimental features on extra branches. Usually one branch for one new feature. Something like development, but try independent features independent, so whichever is ready can be merged to master. If they depend on each other, merge them or rebase against dependent branches. Merge them once they are stable enough and somehow tested.  Just tag release on master branch, it should be enough.

I do not think we have to make *perfect* release. Just try it to be usable. Tagging something 2.0a1, 2.0a2, 2.0rc1, and 2.0 release is not much work. We do not have to wait years to make next release if bugs are found.

**Comment by hweimer** [comment]
One other reason against including PBEM in the upcoming release are security vulnerabilities. Loading PBEM saves means loading untrusted data. Unless we want to promote the release by a "marketing by CVE" strategy, we probably want to iron out the worst things first. Adding proper serialization as I mentioned in #21 would help a lot, but there are also some easily exploitable calls to sprintf() around.

**Comment by peyre** [comment]
I get that in principle, but how likely are we to find someone using a BUZZ#.SAV file as a malicious payload,  spoofing the address of the person you happen to be PBEMing with?

**Comment by hweimer** [comment]
Well, at some point it would be cool to have a mechanism where people can announce their interest to play a PBEM game. Whether it should be something external like a mailing list or some kind of multiplayer central integrated into the game is another question, but that would mean opening SAVs from random strangers on the internet.

**Comment by peyre** [comment]
That would be cool.  Another snappy feature would be if the game could email the save file automatically to the address it came from - but that's probably a bit out of scope.

**Comment by rnyoakum** [comment]
> but there are also some easily exploitable calls to sprintf() around

I've tried to eliminate all of those with commit 2d337e5e83a284cd8ab6a0331db0f3c088888d29. I've had that on a TODO list for a while now, so thanks for the reminder. Also a good reminder that the game wasn't written with security in mind.

**Comment by peyre** [comment]
Incidentally, I've been keeping a Changelog.txt file.  I've tried to add everything we've done beyond getting it working as v.1.1 did, classified into bug fixes and game changes.  (I've also been posting them on the Raceintospace twitter feed.)  So when the time comes, I have a changelog ready that we can work from.

**Comment by rnyoakum** [comment]
> I've also been posting them on the Raceintospace twitter feed

Wait... what??? Race Into Space has a twitter feed? Where?

**Comment by peyre** [comment]
Yep, it's @raceintospace.  It shows my real name there, on my feed.  Don't know if it shows that for others.

**Comment by rnyoakum** [comment]
> Regarding Play by Email, I think we can make it optional. It could be enabled or disabled on compile time. So it does not spoil release for those who wants it perfect. But to provide simple way to enable experimental features. Simple use of precompiler define, enabled by CMake boolean EXPERIMENTAL, might help.

I submitted a pull request (https://github.com/raceintospace/raceintospace/pull/480) that would require the CMake flag "PBEM" be set to use the Play-by-Email feature. Otherwise, the compiler flag won't be set and access to the PBEM menu option and PBEM saves will be disabled. It works fine for me under Linux, but if someone could test it on their system I'd appreciate that before it is committed.

**Comment by pemensik** [comment]
I have prepared alpha tag for 2.0.0. Added version into the credits, so it shows in game what version it it. I think it should also include *git describe* output somewhere. but static version should work well for now.

**Comment by peyre** [comment]
So, what do we have left to do before we can put out the 2.0 release?  Are there any remaining showstoppers?  Could we put out a release this weekend or next, perhaps?

**Comment by pemensik** [comment]
I have prepared v2.0.0rc1 tag, and tested builds on [Fedora](https://koji.fedoraproject.org/koji/taskinfo?taskID=55590816). It will keep the files only for 3 days from build date, don't be if they would be deleted already. I think we should try preparing of installers for version 2.0.0, and commit any changes that might be required to internal code. Do we have other packages to test? Installer, which might be tested it creates working game installation?

I would prepare 2.0.0 final release, if there are no important bugs or bugs we can fix soon.

**Comment by peyre** [comment]
Great, I'd like for people to be able to play with the improvements we've made so far.

So what platforms will have binaries?  Windows I assume, and I guess we'll have a Fedora .rpm; that's good.  I would like if possible to also have a .deb installer.  I assume though that Mac is still giving us trouble?

**Comment by pemensik** [comment]
I thought we may make a release on 2020-12-12 date, but due to issue #420 outcome, I think we should wait for clear definition what can be redistributed and under what conditions. It would be important for Linux free software distributions.

**Comment by peyre** [comment]
Is there anything I can do to help prepare the release?  I've never worked on a release before, but I would like to learn (if I can get it through my thick head anyway) and I'm eager to see it happen.

**Comment by peyre** [comment]
Hey folks!  Any chance we could work on it this weekend?  I'm available if there's anything I can do to help.

**Comment by hweimer** [comment]
Concerning, #420, wouldn't it be the easiest option to make a release without the music, as this appears to be the only potentially problematic asset?

**Comment by peyre** [comment]
Given that Fritz seems to be otherwise occupied, that might be a good idea.  Maybe we could keep the in-game stuff and leave off the beginning and end music.  Seems to me that's what's most at issue here.

**Comment by peyre** [comment]
[b]So,[/b] now that #420 is resolved, maybe we should proceed with a release!  Anyone up for it?

Note: we might need to replace the orchestral music with midi files (data/midi/liftoff.mid and victory.mid).

**Comment by pemensik** [comment]
So, I have verified a Flatpak is buildable again with more recent version. Created another [v2.0.0rc2 tag](https://github.com/raceintospace/raceintospace/releases/tag/v2.0.0rc2). Are there any blockers, which should get fixed, until we make official release?

**Comment by peyre** [comment]
I really would like to have the audio bug (#209, #215) fixed, but it shouldn't stand in the way of us putting out a release.

There is one showstopper that I know of.  As much as I'll miss the orchestral music for opening and endgame, Fritz has reserved the copyright on those (#420), so I think we have to replace them with liftoff.mid and victory.mid respectively.

**Comment by peyre** [comment]
So, should I hold off on further development for now?

**Comment by pemensik** [comment]
Oh, I thought issue #420 were fixed enough. Ok, it should be simple enough to replace them.

I think improvements are possible, but please postpone changes with higher chance to break something. I think fixes of issues should now be made.

**Comment by pemensik** [comment]
Oh, I have regenerated them again just to be sure. But it seems Hendrik were right, they seem to sound exactly the same. But that should be a proof they are made from midi only and not from non-released CD track.

**Comment by hweimer** [comment]
The problem with the music files is that even the MIDI files (and hence, the OGGs generated from them) cannot be distributed under the GPL (see the discussion in #573). The best solution would be to move the music into a separate GItHub repository (say "raceintospace/raceintospace-music") and have source releases consisting of two tarballs. One non-GPLed for the music and one GPLed for the rest. Binary releases can still be distributed together, provided there is a clear statement that the music is not licensed under the terms of the GPL.

**Comment by peyre** [comment]
I've downloaded the repository and recompiled.  At first listen, it sounded a little lower-quality than what we had before, which sounded right--but that might have been me _expecting_ it to sound different.

So I had a listen to the opening sound of the floppy and CD versions, and there's a world of difference between those!  (At least, for the floppy version in SoundBlaster running on DOSBox--I haven't tried other audio cards.)  Ours, in fact, sounds right in between them.  It's higher-quality than in the floppy version, but it's also clearly not the orchestral track from the CD version.  I think we're ok.

Floppy intro:  
CD intro:  [https://www.youtube.com/watch?v=sUIWUNzyVTI](url)

**Comment by peyre** [comment]
So, is this release something we can put together soon?

**Comment by peyre** [comment]
Sorry Petr, I think I was mistaken--that the game was already using the MIDI files rather than orchestral versions.  I think the endgame music was better before the regeneration, and the intro music as well.  Can we go back to what we had?

**Comment by peyre** [comment]
Hey, it looks like we have Fritz's blessing to proceed!  #672

**Comment by hweimer** [comment]
With #720 being merged, I think we can enable PBEM in the upcoming release. @peyre and I did some extensive playtesting, and it looks like we have squashed most serious bugs.

**Comment by peyre** [comment]
I second that.  I'm seeing some occasional segfaults during missions (#687), and it would've been nice if we could've fixed the problem with the music (#209), but I'm inclined to think we're ready for 2.0.

**Comment by peyre** [comment]
Now that we've turned the new year, it occurs to me that this July 4 will be the 10th anniversary of the official release of 1.1.  Maybe we could make it a priority to get the 2.0 release out before then?
